### PR TITLE
Use __cplusplus to test the existence of std::span

### PR DESCRIPTION
**Pull request by zhao-shihan**
# This Pull request:

## Changes or fixes:

When the external application (depending on ROOT) uses C++20, adding this C++ feature test can reduce the possibility of compilation failure.

## Checklist:

- [x] tested changes locally

### CODE REVIEW

**Comment by pcanal** [comment]
In what situation is `R__HAS_STD_SPAN` not sufficient?  If `ROOT` has been configured to build in C++20 mode and properly installed, it must be set in `RConfigure.h`.  If it is not set, this means that the `ROOT` code will use an alternative to `std::span` (because at configure and build time, it detected it was not available).

**Comment by zhaoShiEnthalpy** [comment]
> In what situation is `R__HAS_STD_SPAN` not sufficient? If `ROOT` has been configured to build in C++20 mode and properly installed, it must be set in `RConfigure.h`. If it is not set, this means that the `ROOT` code will use an alternative to `std::span` (because at configure and build time, it detected it was not available).

In fact, I have a project that uses C++20, and uses ROOT compiled with C++17, and I encountered compilation errors in the ROOT header file RSpan.hxx as mentioned (redefinition of std::span). However, once the restriction of this preprocessor macro is resolved, the compilation could pass, and no related problems have been encountered for the time being. I also checked RStringView.hxx, this header file also uses a similar technique, so I submitted this patch. I believe that I may not be considerate about this, but I hope that the issues of "mixing standards", which seems not to be issues, can be solved. If there are indeed many restrictions/issues on this issue, I admit that this is not such simple...

**Comment by pcanal** [comment]
> In fact, I have a project that uses C++20, and uses ROOT compiled with C++17,

We currently do not support this mixing of standard.

> However, once the restriction of this preprocessor macro is resolved, the compilation could pass, and no related problems have been encountered for the time being. 

This means that the example/test that have been run so far did not stress the interfaces that uses `std::span` (and/or your lucky that both version are similar enough).  In practice, changing the test as this PR does introduces a ODR (one definition rule) violation: i.e. some of the code use one version of `std::span` and some of the code use a different version of `std::span`

**Comment by zhaoShiEnthalpy** [comment]
> In practice, changing the test as this PR does introduces a ODR (one definition rule) violation: i.e. some of the code use one version of `std::span` and some of the code use a different version of `std::span`

Exactly, that's the tricky part of this issue. If no other means are used to avoid problems, and the compilation is successful, directly adding the same name to the std will potentially lead to many surprises. 

If I understand correctly, the same is true for `std::string_view`. See `RStringView.hxx`:

```C++
#if defined(R__HAS_STD_STRING_VIEW) || R__MSVC_LANG >= 201703L || __cplusplus >= 201703L

#include <string_view>

#else

# if defined(R__HAS_STD_EXPERIMENTAL_STRING_VIEW)
#  include <experimental/string_view>
# else
#  include "RWrap_libcpp_string_view.h"
# endif
```

There is also a feature test macro. If someone uses C++17 and links to the ROOT compiled by C++14, they might encounter the same issue.

I think it is a better practice to change a namespace (such as `cxx20::span`, `cxx17::string_view`, etc.) and provide the corresponding implicit conversion when the standard matches. In this way, users will not be disturbed to use normally, and these features are also available inside ROOT. The only symmetry break is in the API. Because the features exposed are not those of the standard library, users may need to manually convert them in some cases, so implicit conversion should be provided. 

For example:

```C++
#if __cplusplus >= 202002L
#  include <span>
#endif

namespace cxx20 {

constexpr std::size_t dynamic_extent =
#if __cplusplus >= 202002L
    std::dynamic_extent;
#else
    /* something */;
#endif

template<class T, std::size_t Extent = dynamic_extent>
class span {

// ... here is the normal implementation

#if __cplusplus >= 202002L
public:
    constexpr operator std::span<T, Extent>() noexcept;
    constexpr span(const std::span<T, Extent>& other) noexcept;
    constexpr span& operator=(const std::span<T, Extent>& other) noexcept;
#endif

// ... here is the normal implementation, too

};

}
```

It seems that sometimes ROOT programming does not strictly follow the C++ standard... For me, it is not reassuring to rely on the well-defined situation of undefined behaviors. It is really tricky to handle.

**Comment by Axel-Naumann** [comment]
We require that the language standard including the header is the same as the one used to configure ROOT. Would enforcing that be another option to solve this?

Can you point me to documentation of GCC stating that mixing language standards across libraries is supported?

**Comment by zhaoShiEnthalpy** [comment]
If I understand correctly, this basically involves ABI compatibility. According to the GCC [document (ABI policy)](https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html), in C++, ABI is a mixture of standard library API and compiler ABI. Suppose there are two ideal libraries libA and libB (that is, they are ABI compatible except for the standard library), as long as they are generated by the common standard library API and compiler ABI, they are ABI compatible.

From this point of view, as long as users use the same compiler (or at least compilers with the same compiler ABI) to compile libA and libB (this is a common behavior), the compiler ABI condition is automatically met. On the other hand, the C++ standard library API is specified by ISO C++, and we can believe that the implementation of the standard library is compatible between different standards. So generally speaking, libA and libB are compatible even if they are compiled by different standards.

However, in practice, the sequence of tokens of the same entity in the library may be different between different C++ standards, so the same symbol (such as a function) may correspond to different binaries, which indeed violates the ODR. However, as long as their behaviors are compatible, no matter which version of the linker is selected, there should be no serious problems. In other words, it can be considered that this is only a minor ODR violation, because it does not produce observable effects.

The GCC document does not seem to clearly indicate whether compiling with mixed standards would work, but its [ABI policy](https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html) seems to imply that this is feasible.

In addition, this issue has also been [discussed on StackOverflow](https://stackoverflow.com/questions/46746878/is-it-safe-to-link-c17-c14-and-c11-objects), and the view is that it is feasible to do so.

Back to our current issue, ideally, as long as there is no explicit ODR violation in the header file, the only thing left is to ensure ABI compatibility, which should be guaranteed by the compiler.

It can be said that if the current `span` implementation of ROOT is compatible with the implementation of the standard library, the problem of mixed standards is not serious, but it is difficult to always guarantee. But as I said, if there is something like `cxx20::span` instead of crashing with the standard library, this issue can be solved from its source. The rest is just the ABI compatibility of this `cxx20::span`, which is guaranteed by the compiler.

**Comment by Axel-Naumann** [comment]
I cannot follow. Can you point out where GCC states that their ABI and stdlib API is compatible for different C++ standard versions? To my knowledge that is not guaranteed, for instance mangling has changed between standard versions.

**Comment by zhaoShiEnthalpy** [comment]
> ... this is only the case once the support is deemed stable. To my knowledge even this is not guaranteed, for instance mangling has changed between standard versions.

As for mangling (as part of ABI): as long as the API of the standard library and the ABI used by the compiler remain unchanged, the final mangling is the same. For the latter, the compiler ABI is the same on the same compiler, so the mangling rules are the same. Therefore, as long as the API of the standard library is stable between standards, ABI compatibility can be guaranteed. In my opinion, this should be guaranteed for the standard that has been stabilized (for example, conservatively, no matter C++11 or C++17, the mangling of the same entity should be the same).

> Can you point out where GCC states that their ABI and stdlib API is compatible for different C++ standard versions?

So far, I have not found any direct description of standard mixing in GCC documents, but it seems that they do not point out that changing standards will lead to changes in ABI compatibility (If so, it means that the backward compatibility of the C++ standard is violated, which seems serious). I tend to think that this compatibility is guaranteed, but it does require more evidence.

> I also don't understand how `cxx20::span` would solve anything: we want users to be able to use the stdlib `span` with ROOT, if their stdlib provides it.

In fact, if the user uses the same standard as ROOT, the current situation is very good.

I propose to change the name on the premise that users are allowed to link to ROOT with different standards. Typically, if the user is using a higher standard, the `std::span` in the std library will be exposed and conflict with the ROOT version `std::(inline)__ROOT::span`. In order to resolve the conflict, we have to take the second place: if we use `cxx20::span` and provide an interface to convert from/to standard `std::span`, then even if there are differences in standards, the two sides will not interfere with each other. If users have `std::span`, they can use it; If not, they should not get an implementation from ROOT in the std namespace (which affects their subsequent upgrade to a higher version of the C++ standard). All in all, it provides an implementation independent of std when the true `std::span` is not available, and prevents name conflict when mixing the standard.
Of course, if want ROOT to use `std::span` of the standard library when feasible (e.g. complie with C++20), we can just use `cxx20::span` as the alias of `std::span`. However, after then the user cannot be allowed to go back to a standard lower than C++20, otherwise `span` will missing in the header file, and the ROOT internal implementation of `span` should not be use as the fallback - if so, there may be a compatibility problem, caused by the so-called ODR violation.

> To come back to my question: should we just forbid compiling ROOT headers with a standard that doesn't match what ROOT was built with?

My opinion is to allow this, even if it is not recommended. As mentioned above, issues result from cross standards are mainly caused by changes in the header API. If ROOT has confidence in the compatibility of header files across standards, I don't think it is necessary to worry about mixing standards. If there are still some potential issues, I think it is beneficial to 
firstly ensure successful compilation when mixing standards.

**Comment by Axel-Naumann** [comment]
> I propose to change the name on the premise that users are allowed to link to ROOT with different standards.

That's a misconception; they are not allowed to do that.

We prefer that people can use the stdlib types (including their documentation) and thus intentionally use the same spelling. ROOT's "backports" are really only meant to exist for cases where the stdlib does not provide these types; mixing language versions does obviously not work with that.

**Comment by zhaoShiEnthalpy** [comment]
Anyway, back to this PR: the motivation is the same as https://github.com/root-project/root/commit/ff95b036a75e9c9b33223907ba32862dce0e55f7. At least I want to confirm whether it is appropriate.

File: core/foundation/inc/ROOT/RSpan.hxx [hunk 1]
```diff
@@ -16,7 +16,13 @@
...
 #include "RConfigure.h"
 
-#ifdef R__HAS_STD_SPAN
+#if defined(_MSVC_LANG)
+# define R__MSVC_LANG _MSVC_LANG
+#else
+# define R__MSVC_LANG 0
+#endif
+
+#if defined(R__HAS_STD_SPAN) || R__MSVC_LANG >= 202002L || __cplusplus >= 202002L
```
Comment (Axel-Naumann) [inline core/foundation/inc/ROOT/RSpan.hxx:25]:
```suggestion
#if defined(R__HAS_STD_SPAN) || _MSVC_LANG >= 202002L || __cplusplus >= 202002L
```

**Comment by Axel-Naumann** [comment]
We don't want to change the spelling of the std types. The implementation can be in whatever namespace as long as it's different from the stdlib namespace (to cause linker errors if `std::foo` is passed to a ROOT-backport `std::foo`). I suppose that's still the case here, so: all good!

**Review by Axel-Naumann — CHANGES_REQUESTED** [review]
We don't need the preproc var fallback:
 
> After all replacements due to macro expansion and the defined unary operator have been performed, all remaining identifiers and keywords, except for true and false, are replaced with the pp-number 0

File: core/foundation/inc/ROOT/RSpan.hxx (hunk 1 above)
Comment (zhaoShiEnthalpy) [reply core/foundation/inc/ROOT/RSpan.hxx]: Great, then `_MSVC_LANG` is enough. Updated.

**Review by Axel-Naumann — APPROVED** [review]
Thanks!

**Comment by Axel-Naumann** [comment]
@phsft-bot build

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on windows10/cxx14

**Comment by zhao-shihan** [comment]
CI passed.

**Comment by Axel-Naumann** [comment]
Apologies for letting this sit way too long. I think what we need is:
- get rid of `R__HAS_STD_SPAN` (incl the CMake code)
- remove the option of `experimental::span` (incl CMake) as I don't manage to find any incarnation of this anyway.
- use the feature test macro `__cpp_lib_span`

I would totally understand if you'd find this unreasonable at this point. Please let me know and I will close this PR and implement myself what I suggest here.

**Review by Axel-Naumann — CHANGES_REQUESTED** [review]
See https://github.com/root-project/root/pull/11311#issuecomment-1344293553

**Comment by Axel-Naumann** [comment]
Turns out I *need* to fix this for C++20 support in ROOT - I'm on it.

**Comment by Axel-Naumann** [comment]
That's (currently) [`4ac5bb4` (#11874)](https://github.com/root-project/root/pull/11874/commits/4ac5bb437e47ad8f6ccc9fc21778a56bcfd4d257)

**Comment by Axel-Naumann** [comment]
Closing this now that https://github.com/root-project/root/pull/11874 is merged. Note that we're not there yet; some parts for full C++20 support are still missing. I expect that 6.28/04 will finally provide C++20 support.

**Comment by zhao-shihan** [comment]
> Closing this now that #11874 is merged. Note that we're not there yet; some parts for full C++20 support are still missing. I expect that 6.28/04 will finally provide C++20 support.

Thanks!

**Pull request closed without merging**